### ISSUE TITLE
Deploy fails on merge conflict after force-push

### ISSUE THREAD

**Issue by Gaen** [Unitech/pm2#4390]
## What's going wrong?

1. Have some code in the remote repository
2. Deploy it (works fine)
3. Force-push some conflicting changes
4. Deploy again

This results in deploy failing because of a merge conflict. Output looks like this:
```
--> Deploying to sandbox environment
--> on host localhost
  ○ deploying origin/master
  ○ executing pre-deploy-local
  ○ hook pre-deploy
  ○ fast forward master
Already on 'master'
Your branch is up to date with 'origin/master'.
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0)
Unpacking objects: 100% (3/3), done.
From /Users/gaen/dev/pm2-force/dev/../remote
 * branch            master     -> FETCH_HEAD
 + 8800e4a...8b96829 master     -> origin/master  (forced update)
Auto-merging sample.js
CONFLICT (content): Merge conflict in sample.js
Automatic merge failed; fix conflicts and then commit the result.

  git pull failed

Deploy failed

```

One might expect pm2 to perform `git reset --hard` and succeed with deploy, at least when being run with `--force` flag.

## How could we reproduce this issue?

Here is a script that reproduces the issue from scratch. It assumes that `pm2` is installed globally. 

```bash
# Preclean - use with caution
#rm -rf remote.git
#rm -rf dev
#rm -rf /tmp/pm2sandbox
#pm2 -s del sample

#----------------------------------------------------------------------

# Init remote repo
mkdir remote.git
git init -q --bare remote.git

#----------------------------------------------------------------------

# Init dev repo
mkdir dev
git clone -q remote.git dev &> /dev/null

#----------------------------------------------------------------------

# Enter dev repo
pushd dev > /dev/null

# Create initial version of the file
cat > sample.js <<EOF
console.log('Hi!');
console.log('Bye!');
EOF

# Create deploy config
cat > ecosystem.json <<EOF
{
   "apps" : [{
      "name" : "sample",
      "script" : "sample.js"
   }],
   "deploy" : {
     "sandbox" : {
       "host" : ["localhost"],
       "ref"  : "origin/master",
       "repo" : "../remote.git",
       "path" : "/tmp/pm2sandbox",
      }
   }
}
EOF

# Commit and push
git add . && git commit -q -m "initial commit" && git push -q

#----------------------------------------------------------------------

# Set up deploy environment
echo "##### Setting up deploy environment"
pm2 deploy sandbox setup #&> /dev/null

# Perform initial deploy
echo "##### Performing initial deploy"
pm2 deploy sandbox #&> /dev/null

#----------------------------------------------------------------------

# Make some changes
cat > sample.js <<EOF
console.log('Hi!');
console.log('Deploy automation is a joy!');
console.log('Bye!');
EOF

# Commit and push
git add . && git commit -q -m "some changes" && git push -q

# Deploy changes
echo "##### Deploying regular changes"
pm2 deploy sandbox #&> /dev/null

#----------------------------------------------------------------------

# Introduce conflicting changes
cat > sample.js <<EOF
console.log('Hi!');
console.log('Deploy automation is a pain!');
console.log('Bye!');
EOF

# Amend and force-push
git add . && git commit -q --amend -m "some conflicting changes" && git push -f -q

# Ensure default git pull behaviour for deploy environment
pm2 deploy sandbox exec git config pull.rebase false > /dev/null

# Deploy changes
echo "##### Deploying conflicting changes"
pm2 deploy sandbox

#----------------------------------------------------------------------

# Exit dev repo
popd > /dev/null
```

## Supporting information

Tested under the following conditions:
```bash
# macOS Mojave v10.14 on iMac 2017

$ git --version
git version 2.22.0

$ node --version
v10.14.0

$ pm2 --version
3.5.1
```

`pm2 report` output as requested in issue template:

```
--- PM2 report ----------------------------------------------------------------
Date                 : Tue Aug 13 2019 19:50:00 GMT+0300 (Moscow Standard Time)
===============================================================================
--- Daemon -------------------------------------------------
pm2d version         : 3.5.1
node version         : 10.14.0
node path            : /Users/gaen/.nvm/versions/node/v10.14.0/bin/pm2
argv                 : /Users/gaen/.nvm/versions/node/v10.14.0/bin/node,/Users/gaen/.nvm/versions/node/v10.14.0/lib/node_modules/pm2/lib/Daemon.js
argv0                : node
user                 : gaen
uid                  : 502
gid                  : 20
uptime               : 44min
===============================================================================
--- CLI ----------------------------------------------------
local pm2            : 3.5.1
node version         : 10.14.0
node path            : /Users/gaen/.nvm/versions/node/v10.14.0/bin/pm2
argv                 : /Users/gaen/.nvm/versions/node/v10.14.0/bin/node,/Users/gaen/.nvm/versions/node/v10.14.0/bin/pm2,report
argv0                : node
user                 : gaen
uid                  : 502
gid                  : 20
===============================================================================
--- System info --------------------------------------------
arch                 : x64
platform             : darwin
type                 : Darwin
cpus                 : Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
cpus nb              : 8
freemem              : 21884928
totalmem             : 8589934592
home                 : /Users/gaen
===============================================================================
--- PM2 list -----------------------------------------------
┌──────────┬────┬─────────┬──────┬─────┬─────────┬─────────┬────────┬─────┬────────┬──────┬──────────┐
│ App name │ id │ version │ mode │ pid │ status  │ restart │ uptime │ cpu │ mem    │ user │ watching │
├──────────┼────┼─────────┼──────┼─────┼─────────┼─────────┼────────┼─────┼────────┼──────┼──────────┤
│ sample   │ 0  │ N/A     │ fork │ 0   │ errored │ 16      │ 0      │ 0%  │ 0 B    │ gaen │ disabled │
└──────────┴────┴─────────┴──────┴─────┴─────────┴─────────┴────────┴─────┴────────┴──────┴──────────┘
===============================================================================
--- Daemon logs --------------------------------------------
/Users/gaen/.pm2/pm2.log last 20 lines:
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] exited with code [1] via signal [SIGINT]
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] starting in -fork mode-
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] online
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] exited with code [1] via signal [SIGINT]
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] starting in -fork mode-
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] online
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] exited with code [1] via signal [SIGINT]
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] starting in -fork mode-
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] online
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] exited with code [1] via signal [SIGINT]
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] starting in -fork mode-
PM2        | 2019-08-13T19:14:44: PM2 log: App [sample:0] online
PM2        | 2019-08-13T19:14:45: PM2 log: App [sample:0] exited with code [1] via signal [SIGINT]
PM2        | 2019-08-13T19:14:45: PM2 log: App [sample:0] starting in -fork mode-
PM2        | 2019-08-13T19:14:45: PM2 log: App [sample:0] online
PM2        | 2019-08-13T19:14:45: PM2 log: App [sample:0] exited with code [1] via signal [SIGINT]
PM2        | 2019-08-13T19:14:45: PM2 log: App [sample:0] starting in -fork mode-
PM2        | 2019-08-13T19:14:45: PM2 log: App [sample:0] online
PM2        | 2019-08-13T19:14:45: PM2 log: App [sample:0] exited with code [1] via signal [SIGINT]
PM2        | 2019-08-13T19:14:45: PM2 log: Script /private/tmp/pm2sandbox/source/sample.js had too many unstable restarts (16). Stopped. "errored"

```

**Comment by charlieclark** [comment]
I also just encountered this - would be great if there was a way to deploy changes that were force-pushed.

**Comment by cooperka** [comment]
To be clear, this can be resolved on the machine you're trying to deploy to via:
1. `cd DEPLOY_LOCATION/source/`
1. `git reset --hard origin/BRANCH`

Frustrating that there's seemingly no option to force deploys like this automatically, but at least it's a simple workaround.

**Comment by Gaen** [comment]
A bit simpler workaround, useful when there are multiple target machines:

`pm2 deploy <environment> exec git reset --hard origin/<branch>`

**Comment by AoDev** [comment]
@Gaen Hi Gaen. When trying your work around, I get _error: unknown option `--hard'_. Am I missing something? 

My full command looks like this: `pm2 deploy ecosystem.config.local.yml game-development exec git reset --hard origin/dev`

**Comment by prusswan** [comment]
I had to enclose the git command in quotes:

> pm2 deploy <environment> exec "git reset --hard origin/`<branch>`"

**Comment by Gaen** [comment]
My bad, missed the quotes. Updated the workaround command.

I've also added `git fetch` there. It helps in case you're trying to deploy a branch which has been hard-reset to a previous commit and then force-pushed.

**Comment by AoDev** [comment]
There is an undocumented [`pre-deploy` hook](https://github.com/Unitech/pm2/issues/4687).
I have my config like this: 

```
  some-deployment:
    ref: origin/dev
    repo: ...
    path: ...
    pre-deploy: "git fetch && git reset --hard origin/dev"
```

**Comment by Gaen** [comment]
Hm it looks like this issue has been addressed in `pm2@4.3.0` which has been released on 2020-04-18. Checked with script from op-post, it doesn't fail anymore.

Changelog mentions this as "chore: pm2-deploy@1.0.2 (revert fix)". Actual commit is here: https://github.com/Unitech/pm2-deploy/commit/987428a6ce402b47bb195ab93a8eafa5ad79c6f5
